renderdoc/driver/vulkan/vk_acceleration_structure.cpp

```diff
@@ -580,183 +613,265 @@ bool VulkanAccelerationStructureManager::Serialise(SerialiserType &ser, Resource
+  if(asInfo->replayAS == VK_NULL_HANDLE)
   {
-    const VkDeviceSize size =
-        AlignUp(initial.mem.size, m_pDriver->GetDeviceProps().limits.nonCoherentAtomSize);
+    rdcarray<VkAccelerationStructureBuildRangeInfoKHR> buildRangeInfos = asInfo->getBuildRanges();
+    rdcarray<VkAccelerationStructureGeometryKHR> asGeomData = asInfo->convertGeometryData();
+    RDCASSERT(!asGeomData.empty());
+    RDCASSERT(asInfo->geometryData.size() == asGeomData.size());
+
+    if(!FixUpReplayBDAs(asInfo, asGeomData))
+      return;
+
+    // Allocate the scratch buffer which involves working out how big it should be
+    VkAccelerationStructureBuildSizesInfoKHR sizeResult = {
+        VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_BUILD_SIZES_INFO_KHR,
+    };
+    {
+      const VkAccelerationStructureBuildGeometryInfoKHR sizeInfo = {
+          VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_BUILD_GEOMETRY_INFO_KHR,
+          NULL,
+          asInfo->type,
+          asInfo->flags,
+          VK_BUILD_ACCELERATION_STRUCTURE_MODE_BUILD_KHR,
+          VK_NULL_HANDLE,
+          VK_NULL_HANDLE,
+          (uint32_t)asGeomData.size(),
+          asGeomData.data(),
+          VK_NULL_HANDLE,
+      };
+
+      rdcarray<uint32_t> counts;
+      counts.reserve(asGeomData.size());
+      for(VkAccelerationStructureBuildRangeInfoKHR numPrims : buildRangeInfos)
+        counts.push_back(numPrims.primitiveCount);
+
+      ObjDisp(d)->GetAccelerationStructureBuildSizesKHR(
+          Unwrap(d), VK_ACCELERATION_STRUCTURE_BUILD_TYPE_DEVICE_KHR, &sizeInfo, counts.data(),
+          &sizeResult);
+    }
+    UpdateScratch(sizeResult.buildScratchSize);
+
+    cmd = m_pDriver->GetInitStateCmd();
+    if(cmd == VK_NULL_HANDLE)
+    {
+      RDCERR("Couldn't acquire command buffer");
+      return;
+    }
+
+    // Create the base AS
+    const VkBufferCreateInfo gpuBufInfo = {
+        VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
+        NULL,
+        0,
+        sizeResult.accelerationStructureSize,
+        VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_STORAGE_BIT_KHR |
+            VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT_KHR,
+    };
+
+    VkBuffer asBuf = VK_NULL_HANDLE;
+    VkResult vkr = m_pDriver->vkCreateBuffer(d, &gpuBufInfo, NULL, &asBuf);
+    CHECK_VKR(m_pDriver, vkr);
+
+    VkMemoryRequirements mrq = {};
+    ObjDisp(d)->GetBufferMemoryRequirements(Unwrap(d), Unwrap(asBuf), &mrq);
+    mrq.alignment = AlignUp(mrq.alignment, asBufferAlignment);
 
-    // Copy the data using host-commands but from mapped memory
-    byte *mappedSrcBuffer = NULL;
-    VkResult vkr = ObjDisp(d)->MapMemory(Unwrap(d), Unwrap(initial.mem.mem), initial.mem.offs, size,
-                                         0, (void **)&mappedSrcBuffer);
+    const MemoryAllocation asMemory = m_pDriver->AllocateMemoryForResource(
+        true, mrq, MemoryScope::InitialContents, MemoryType::GPULocal);
+    vkr = m_pDriver->vkBindBufferMemory(d, asBuf, asMemory.mem, asMemory.offs);
```

Comment:
I think we need to hold onto this `asBuf` and `asMemory` somewhere so we can properly destroy them once we're done with the initial contents at replay shutdown? At the moment we only hold onto `replayAS`, and  destroy the upload buffer/memory.

Reply:
`asBuf` yes, but won't `asMemory` be cleaned up when all the Initial contents are?

Reply:
Ah yes sorry it will, ignore me reading too quickly :) only `asBuf` indeed.